sasl.go

```diff
@@ -88,3 +94,85 @@ func ConnSASLAnonymous() ConnOption {
 		return nil
 	}
 }
+
+// ConnSASLXOAUTH2 enables SASL XOAUTH2 authentication for the connection.
+func ConnSASLXOAUTH2(username, bearer string, saslMaxFrameSizeOverride uint32) ConnOption {
+	return func(c *conn) error {
+		// make handlers map if no other mechanism has
+		if c.saslHandlers == nil {
+			c.saslHandlers = make(map[symbol]stateFunc)
+		}
+
+		response, err := saslXOAUTH2InitialResponse(username, bearer)
+		if err != nil {
+			return err
+		}
+
+		// add the handler the the map
+		c.saslHandlers[saslMechanismXOAUTH2] = func() stateFunc {
+			originalPeerMaxFrameSize := c.peerMaxFrameSize
+			if saslMaxFrameSizeOverride > c.peerMaxFrameSize {
+				c.peerMaxFrameSize = saslMaxFrameSizeOverride
+			}
+			c.err = c.writeFrame(frame{
+				type_: frameTypeSASL,
+				body: &saslInit{
+					Mechanism:       saslMechanismXOAUTH2,
+					InitialResponse: response,
+				},
+			})
+			c.peerMaxFrameSize = originalPeerMaxFrameSize
+			if c.err != nil {
+				return nil
+			}
+
+			// go to c.saslXOAUTH2Step to handle the server response
+			return c.saslXOAUTH2Step
+		}
+		return nil
+	}
+}
+
+func saslXOAUTH2InitialResponse(username string, bearer string) ([]byte, error) {
+	re := regexp.MustCompile("^[\x20-\x7E]+$")
+	if !re.MatchString(bearer) {
+		return []byte{}, fmt.Errorf("unacceptable bearer token")
+	}
```

Comment:
Does the username need to be validated?

Reply:
I could find no authoritative source describing the permitted contents of the username field.  However, I added a restriction disallowing \x01 (as this would interfere with the encoding) and supporting tests.  Also I see use-cases where the user's username is unknown to the caller, so I chose to accept an empty username too.